rfcs/0049-flakes.md

```diff
@@ -0,0 +1,709 @@
+over the local and global registries.
+
+## `nix` command line interface
+
+I propose to make flakes the primary way to specify packages in the
+`nix` command line interface. Note that the `nix` UI is still marked
+as experimental, so we have some freedom to make incompatible
+changes. The legacy commands (`nix-build`, `nix-shell`, `nix-env` and
+`nix-instantiate`) should not be changed to avoid breakage.
+
+Most `nix` subcommands work on a list of arguments called
+"installables" for lack of a better word. For example,
+
+    # nix run nixpkgs:hello dwarffs:dwarffs
+
+takes two flake-based installables. The general form is:
+
+    <flake-ref>(:<attr-path>)?
+
+Examples of installables:
+
+* `nixpkgs:packages.hello`
+* `nixpkgs:hello` - short for `nixpkgs:packages.hello`
+* `nixpkgs/release-19.03:hello` - overrides the Git branch to use
+* `github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082:hello` -
+  specifies the exact Git revision to use
+* `dwarffs` - short for `dwarffs:defaultPackage`
+* `nix:hydraJobs.build.x86_64-darwin`
+* `.:hydraJobs.build.x86_64-darwin` - refers to the flake in the
+  current directory (which can be a dirty Git tree)
+* `.` - short for `.:defaultPackage`
+
+If no argument is given, the default is `.`; thus,
+
+    # nix build
+
+is equivalent to
+
+    # nix build .:defaultPackage
+
+For backwards compatibility, it's possible to use non-flake Nix
+expressions using `-f`, e.g. `nix build -f foo.nix foo.bar`.
+
+## Lock files and reproducible evaluation
+
+Inputs specified in `flake.nix` are typically "unlocked": they are
+flake references that don't specify an exact revision (e.g. `nixpkgs`
+rather than
+`github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082`). To
+ensure reproducibility, Nix will automatically generate and use a
+*lock file* called `flake.lock` in the flake's directory. The lock
+file contains a tree of mappings from the flake references specified
+in `flake.nix` to direct flake references that contain revisions.
+
+For example, if `flake.nix` contains
+```
+inputs =
+  [ "nixpkgs"
+    github:edolstra/import-cargo
+  ];
+```
+then the resulting lock file might be:
+```
+{
+    "version": 2,
+    "inputs": {
+        "github:edolstra/import-cargo": {
+            "id": "import-cargo",
+            "inputs": {},
+            "narHash": "sha256-mxwKMDFOrhjrBQhIWwwm8mmEugyx/oVlvBH1CKxchlw=",
+            "uri": "github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e"
+        },
+        "nixpkgs": {
+            "id": "nixpkgs",
+            "inputs": {},
+            "narHash": "sha256-p7UqhvhwS5MZfqUbLbFm+nfG/SMJrgpNXxWpRMFif8c=",
+            "uri": "github:NixOS/nixpkgs/4a7047c6e93e8480eb4ca7fd1fd5a2aa457d9082"
+        }
+    }
+}
+```
+
+Thus, when we build this flake, the input `nixpkgs` is mapped to
+`github:edolstra/import-cargo/c33e13881386931038d46a7aca4c9561144d582e`. Nix
+will also check that the content hash of the input is equal to the one
```

Comment:
I’m guessing that’s the `narHash`? Wasn’t entirely clear from the paragraph above

Reply:
No, it's the Git revision.